src/OpenTelemetry.Contrib.Instrumentation.GrpcCore/ClientTracingInterceptor.cs

```diff
@@ -0,0 +1,333 @@
+            private readonly ClientInterceptorContext<TRequest, TResponse> context;
+
+            /// <summary>
+            /// The parent activity.
+            /// </summary>
+            private readonly Activity parentActivity;
+
+            /// <summary>
+            /// Initializes a new instance of the <see cref="ClientRpcScope{TRequest, TResponse}" /> class.
+            /// </summary>
+            /// <param name="context">The context.</param>
+            /// <param name="options">The options.</param>
+            public ClientRpcScope(ClientInterceptorContext<TRequest, TResponse> context, ClientTracingInterceptorOptions options)
+                : base(context.Method?.FullName, options.RecordMessageEvents)
+            {
+                this.context = context;
+
+                // Capture the current activity.
+                this.parentActivity = Activity.Current;
+
+                // Short-circuit if nobody is listening
+                if (!GrpcCoreInstrumentation.ActivitySource.HasListeners())
+                {
+                    return;
+                }
+
+                // We want to start an activity but don't activate it.
+                // After calling StartActivity, Activity.Current will be the new Activity.
+                // This scope is created synchronously before the RPC invocation starts and so this new Activity will overwrite
+                // the callers current Activity which isn't what we want. We need to restore the original immediately after doing this.
+                // If this call happened after some kind of async context await then a restore wouldn't be necessary.
+                // gRPC Core just doesn't have the hooks to do this as far as I can tell.
+                var rpcActivity = GrpcCoreInstrumentation.ActivitySource.StartActivity(
+                    this.FullServiceName,
+                    ActivityKind.Client);
+
+                if (rpcActivity == null)
+                {
+                    return;
+                }
+
+                var callOptions = context.Options;
+
+                if (callOptions.Headers == null)
+                {
+                    callOptions = callOptions.WithHeaders(new Metadata());
+                }
+
+                this.SetActivity(rpcActivity);
+                options.Propagator.Inject(new PropagationContext(rpcActivity.Context, Baggage.Current), callOptions.Headers, MetadataSetter);
+                this.context = new ClientInterceptorContext<TRequest, TResponse>(context.Method, context.Host, callOptions);
+            }
+
+            /// <summary>
+            /// Gets the context.
+            /// </summary>
+            public ClientInterceptorContext<TRequest, TResponse> Context => this.context;
+
+            /// <summary>
+            /// Restores the parent activity.
+            /// </summary>
+            public void RestoreParentActivity()
+            {
+                Activity.Current = this.parentActivity;

```

Comment:
I don't think you need this? When you call activity.Stop it should automatically restore the parent. Maybe this is a workaround for https://github.com/dotnet/runtime/pull/48034?

Reply:
So this is actually very necessary. When a call to this made, we are returning from the RPC method itself but the call hasn't actually completed. At the point we return, we need to ensure that Activity.Current is the same as it was when the RPC was invoked otherwise the Activity created by the interceptor bleeds back into the callers which isn't what we want.

Reply:
Interesting. So we're starting the activity on calling thread, then a new thread is started where the rpc happens, calling thread returns and awaits the rpc operation whenever it needs the result? I think the only way to smooth it out would be instead of starting the activity on the calling thread, just grab DateTime.UtcNow and then pass to the rpc thread so it can start the activity using the captured time? Fine by me to leave as is.